x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/pki/PkiRealm.java

```diff
@@ -122,38 +145,45 @@ public void authenticate(AuthenticationToken authToken, ActionListener<Authentic
             User user = cache.get(fingerprint);
             if (user != null) {
                 if (delegatedRealms.hasDelegation()) {
-                    delegatedRealms.resolve(token.principal(), listener);
+                    delegatedRealms.resolve(user.principal(), listener);
                 } else {
                     listener.onResponse(AuthenticationResult.success(user));
                 }
             } else if (isCertificateChainTrusted(trustManager, token, logger) == false) {
                 listener.onResponse(AuthenticationResult.unsuccessful("Certificate for " + token.dn() + " is not trusted", null));
             } else {
-                final ActionListener<AuthenticationResult> cachingListener = ActionListener.wrap(result -> {
-                    if (result.isAuthenticated()) {
-                        try (ReleasableLock ignored = readLock.acquire()) {
-                            cache.put(fingerprint, result.getUser());
+                // parse the principal again after validating the cert chain, and do not rely on the token.principal one, because that could
+                // be set by a different realm that failed trusted chain validation. We SHOULD NOT parse the principal BEFORE this step, but
+                // we do it for BWC purposes. Changing this is a breaking change.
+                final String principal = getPrincipalFromSubjectDN(principalPattern, token, logger);
+                if (principal == null) {
+                    listener.onResponse(AuthenticationResult.unsuccessful("Could not parse principal from Subject DN " + token.dn(), null));
+                } else {
```

Comment:
I wonder if we want some sort of logging if `user.principal()` and `token.principal()` are different.

It might get to noisy to have it as `INFO`/`WARN` every time we do a full-authenticate (potentially every 20 minutes, per user) but maybe a `DEBUG` would help resolve any unexpected behaviours.

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/43932/commits/b16f889bb2ed765e2bd772cdea386b30901edbff which adds debug loggers for when the extracted principal after cert chain validation differs (or is null) from before. These will be useful for us when we need to explain some unexpected behavior, but it is difficult to put in a log line what exactly it's happening so the admin can take actions on its own; so I don't even try, the log lines only state the facts.